Framework/Algorithms/src/XrayAbsorptionCorrection.cpp

```diff
@@ -0,0 +1,232 @@
+ * @param degrees double of angle in degrees
+ * @return A double of angle in radians
+ */
+double XrayAbsorptionCorrection::degreesToRadians(double degrees) {
+  return M_PI * (degrees / 180.0);
+}
+
+/**
+ * Calculates the position of the detector.
+ * @param detectorAngle double of angle of detector
+ * @param detectorDistance distance betweeen sample and detector
+ * @return V3D object of position of detector
+ */
+Kernel::V3D
+XrayAbsorptionCorrection::calculateDetectorPos(double const detectorAngle,
+                                               double detectorDistance) {
+  detectorDistance = detectorDistance * ConversionFrom_cm_to_m;
+  double x = detectorDistance / std::tan(degreesToRadians(detectorAngle));
+  if (detectorAngle > 180.0) {
+    detectorDistance = -detectorDistance;
+  }
+  Kernel::V3D detectorPos = {x, detectorDistance, 0.0};
+  return detectorPos;
+}
+
+/**
+ * normalise moun intensity to 1.
+ * @param muonIntensity A MantidVec which contains the intensity of muons as a
+ * function  of depth
+ * @return A vector of doubles which contains the normalised muon intensity
+ */
+std::vector<double>
+XrayAbsorptionCorrection::normaliseMuonIntensity(MantidVec muonIntensity) {
+
+  double sum_of_elems =
+      std::accumulate(muonIntensity.begin(), muonIntensity.end(), 0.0);
+
+  std::transform(muonIntensity.begin(), muonIntensity.end(),
+                 muonIntensity.begin(),
+                 [sum_of_elems](double d) { return d / sum_of_elems; });
+  return muonIntensity;
+}
+
+/**
+ * Calculate the muon implantation position in the sample.
+ * @param muonProfile A reference to the muon profile workspace
+ * @param inputWS A reference to the input workspace
+ * @param detectorDistance a double representing the distance from the sample to
+ * detector
+ * @return A vector of V3D objects that represent the position of muons
+ */
+std::vector<Kernel::V3D> XrayAbsorptionCorrection::calculateMuonPos(
+    API::MatrixWorkspace_sptr &muonProfile, API::MatrixWorkspace_sptr inputWS,
+    double detectorDistance) {
+  const MantidVec muonDepth = muonProfile->readX(0);
+  Kernel::V3D const muonPoint = {0.0, detectorDistance, 0.0};
+  Kernel::V3D toStart = {0.0, -1.0, 0.0};
+  const Geometry::IObject *shape = &inputWS->sample().getShape();
+  Geometry::Track muonPath = Geometry::Track(muonPoint, toStart);
+  shape->interceptSurface(muonPath);
+  if (muonPath.count() == 0) {
+    throw std::runtime_error("No valid solution, check shape parameters, Muon "
+                             "depth profile and detector distance");
+  }
+  Kernel::V3D intersection = muonPath.cbegin()->entryPoint;
+  double sampleDepth = intersection[1];
+  std::vector<Kernel::V3D> muonPos;
+  for (auto depth : muonDepth) {
+    // Muon implantation position are at x = 0 and z = 0 and y position is
+    // variable
+    Kernel::V3D pos = {0.0, sampleDepth - (depth * ConversionFrom_cm_to_m),
+                       0.0};
```

Comment:
In Mantid, the z axis is normally along the beam with y pointing up. Should the result of this depth calculation be assigned to the z coordinate of pos?